feat(ethexe): Added separate type for network announces

Lazark0x opened the pull request:
# Announce Hash-Only Refactor and `NetworkAnnounce` Wire Payload

## Summary

This refactor splits announce representation into two layers:

1. Local/persisted/compute layer uses `Announce` with:
   - `injected_transactions: Vec<HashOf<InjectedTransaction>>`
2. Network transport layer uses `NetworkAnnounce` with:
   - `injected_transactions: Vec<SignedInjectedTransaction>`

The goal is to stop carrying full injected transaction bodies in every local announce iteration and use hash references locally, while still transferring full signed transactions over the network when needed.

---

## Before vs After

### Before

- `Announce` contained full `Vec<SignedInjectedTransaction>`.
- Announce traversal, tx validity-window scans, and other local logic repeatedly touched heavy tx bodies.
- db-sync announce payloads were local announce objects (full txs embedded).

### After

- `Announce` contains only `Vec<HashOf<InjectedTransaction>>`.
- `NetworkAnnounce` carries full `Vec<SignedInjectedTransaction>` only for wire operations.
- Full tx bodies are persisted in injected-tx storage and fetched by hash only at boundaries that need them.

---

## Type-Level Changes

1. `Announce` field changed in `ethexe/common/src/primitives.rs`:
   - from `Vec<SignedInjectedTransaction>`
   - to `Vec<HashOf<InjectedTransaction>>`

2. `NetworkAnnounce` lives in `ethexe/common/src/primitives.rs`:
   - `block_hash`
   - `parent`
   - `gas_allowance`
   - `injected_transactions: Vec<SignedInjectedTransaction>`
   - and is re-exported from `ethexe/common/src/network.rs` for network-facing API usage.

3. Network aliases switched:
   - `ValidatorAnnounce = ValidatorMessage<NetworkAnnounce>`
   - `VerifiedAnnounce = VerifiedData<NetworkAnnounce>`
   - `AnnouncesResponse` now stores `Vec<NetworkAnnounce>`

4. Conversion APIs on `NetworkAnnounce`:
   - `impl From<&NetworkAnnounce> for Announce`
   - `impl From<NetworkAnnounce> for Announce`
   - `impl TryFrom<(Announce, Vec<SignedInjectedTransaction>)> for NetworkAnnounce`
   - `impl TryFrom<Announce> for NetworkAnnounce` (for hash-only/empty-body test and utility cases)
   - `to_hash()` (hash of converted local `Announce`)
   - `into_announce_persisting_injected_transactions(&db)` to centralize:
     - persisting signed tx bodies to `InjectedStorage`
     - producing local hash-only `Announce`

---

## What Changed for `InjectedTransaction`

## Core rule

`InjectedTransaction` hash is the canonical ID used in local announce chains.

- Hash type: `HashOf<InjectedTransaction>`
- Derived from unsigned payload fields (destination, payload, value, reference_block, salt)
- Not a hash of the signed envelope

## Practical impact

1. Announce chain now stores tx references (hashes), not full signed txs.
2. Signed tx bodies are persisted separately (`InjectedStorage`) keyed by tx hash.
3. Full signed tx body is loaded only when required:
   - network ingress/sync persistence
   - compute executable preparation
4. Missing tx body now has explicit failure paths.

---

## End-to-End Data Flow

## 1) Producer path

```text
Injected pool selection
    |
    v
Vec<SignedInjectedTransaction>
    |\
    | \-- map tx.data().to_hash() --------------------+
    |                                                 |
    v                                                 v
NetworkAnnounce (full tx bodies)                Announce (tx hashes only)
    |                                                 |
    | publish signed validator message                | include in DB
    |                                                 | emit ComputeAnnounce(local)
    v                                                 v
Network                                              Consensus/Compute pipeline
```

## 2) Network announce acceptance path (consensus)

```text
VerifiedAnnounce<NetworkAnnounce> received
    |
    v
accept_announce(network_announce)
    |
    +-- convert to local announce (hash-only, via From impl)
    +-- validate each signed tx from network payload
    +-- persist each valid signed tx:
    |      set_injected_transaction(tx)
    +-- include local announce in announce storage
    |
    v
Emit ComputeAnnounce(local hash-only announce)
```

## 3) Missing-announces sync path

```text
db-sync response: Vec<NetworkAnnounce>
    |
    v
for each network announce:
    network_announce.into_announce_persisting_injected_transactions(db)
      - persist signed tx bodies into InjectedStorage
      - convert to local Announce (hash list)
    pass (announce_hash, announce) to propagation
```

## 4) Compute path

```text
ComputeAnnounce(Announce { tx_hashes... })
    |
    v
prepare_executable_for_announce
    |
    +-- for tx_hash in announce.injected_transactions:
    |      db.injected_transaction(tx_hash) -> SignedInjectedTransaction
    |      into_verified()
    |
    +-- if missing:
    |      ComputeError::InjectedTransactionNotFound(tx_hash)
    |
    v
ExecutableData with full verified txs
```

## 5) db-sync server response path

```text
Local announce in DB (hash-only tx list)
    |
    v
For each tx_hash:
    db.injected_transaction(tx_hash) -> signed tx body
    else -> InjectedTransactionMissing error
    |
    v
Build NetworkAnnounce (full tx bodies) via validated constructor:
    NetworkAnnounce::try_from_announce(announce, injected_transactions)
    |
    v
Send InnerAnnouncesResponse(Vec<NetworkAnnounce>)
```

---

## Hashing and Identity Semantics

There are two relevant hashes now:

1. Local announce hash (`HashOf<Announce>`)
   - computed over local announce fields where tx list is hashes
   - this is canonical for persistence, parent links, and chain checks

2. Signed validator message digest (for transport signing)
   - includes `NetworkAnnounce` payload encoding
   - used for network authenticity, not as local announce chain identity

`NetworkAnnounce::to_hash()` intentionally converts to local announce first (via `From<&NetworkAnnounce>`), then hashes. This keeps network-side validation aligned with local canonical announce hashing.

---

## Storage Boundary (Important)

After the refactor:

- `AnnounceStorage` stores local announce bodies with tx hash references.
- `InjectedStorage` stores full `SignedInjectedTransaction` bodies.

This boundary is now explicit and central:

1. Ingress/sync side writes tx bodies to `InjectedStorage`.
2. Local announce graph carries only hashes.
3. Compute/db-sync server resolve hashes back to full bodies when needed.

---

## Behavior and Error Handling Changes

1. Compute now explicitly errors if announce references a missing tx body:
   - `ComputeError::InjectedTransactionNotFound(HashOf<InjectedTransaction>)`

2. db-sync announce response building explicitly errors if tx hash cannot be resolved:
   - `InjectedTransactionMissing { hash }`
   - and also errors if announce hash-list and loaded tx bodies mismatch:
     - `AnnounceInjectedTransactionsMismatch { hash, source }`
     - where `source` is `NetworkAnnounceFromAnnounceError`

3. DB integrity verifier no longer rejects non-empty announce injected data,
   because non-empty hash lists in announces are valid by design.

---

## Why this is better

1. Better performance for local announce-heavy logic:
   - tx validity-window scans now work on hashes directly.
2. Cleaner data model:
   - announce graph is lightweight and deterministic.
3. Clear separation of concerns:
   - transport payloads vs persisted consensus objects.
4. Explicit failure points:
   - missing tx body surfaces as precise errors instead of implicit assumptions.

---

## Compatibility

This is a breaking storage-format change for announces.

- Old DB data that encoded announces with embedded signed tx bodies is incompatible.
- Resync is required after upgrading.

ecol-master commented on ethexe/common/src/injected.rs:
```diff
@@ -29,6 +29,14 @@ pub const VALIDITY_WINDOW: u8 = 32;
 
 pub type SignedInjectedTransaction = SignedData<InjectedTransaction>;
 
+impl SignedInjectedTransaction {
```
this is not true, use pls to_hash() from InjectedTransaction, because of transaction (and message_id) defines from only its `InjectedTransaction` inner field. 

Signature only for validation and verification.

ecol-master commented on ethexe/common/src/primitives.rs:
```diff
@@ -83,9 +83,7 @@ pub struct Announce {
-    // TODO kuzmindev: remove InjectedTransaction from Announce and store only its hashes.
-    // Need to implement `PublicAnnounce` struct which will contain full bodies of injected transactions.
-    pub injected_transactions: Vec<SignedInjectedTransaction>,
+    pub injected_transactions: Vec<HashOf<SignedInjectedTransaction>>,
```
`Vec<HashOf<InjectedTransaction>>`

Lazark0x committed:
Replace Announce with NetworkAnnounce in network API and related code

Refactor Announce to store only hashes of injected transactions.
Add conversions and mocks for NetworkAnnounce.
Update accept_announce and related functions to use NetworkAnnounce.
Adjust types and imports accordingly throughout the codebase.

Lazark0x committed:
Merge branch 'master' into laz/ethexe/3-blocks/network-announce

Lazark0x committed:
Merge branch 'master' into laz/ethexe/3-blocks/network-announce

Commit:
Merge branch 'master' into laz/ethexe/3-blocks/network-announce

Commit:
Merge branch 'master' into laz/ethexe/3-blocks/network-announce

Commit:
fix: clippy, producer committed logic

kuzminxyz reviewed (commented):
// just to remove `requested changes`

Commit:
fix: accept_announce behavior

Commit:
chore: add check for signature mismatch

Commit:
chore: add test in compute service for injected tx signature mismatch

Commit:
Merge branch 'master' into laz/ethexe/3-blocks/network-announce

kuzminxyz committed:
Merge branch 'master' into laz/ethexe/3-blocks/network-announce

Commit:
note bug: add placeholder for implementing the migration from previous announce value layout to newer

Commit:
Merge branch 'laz/ethexe/3-blocks/network-announce' of https://github.com/gear-tech/gear into laz/ethexe/3-blocks/network-announce

Commit:
feat: draft implementation for network announce database migration

Commit:
Merge branch 'master' into laz/ethexe/3-blocks/network-announce

Commit:
chore: better docs | helper method for NetworkAnnounce

Commit:
chore: remove useless method | fix typo

Commit:
chore: add update config in kv database

kuzminxyz committed:
Merge branch 'master' into laz/ethexe/3-blocks/network-announce

Commit:
chore: refactoring, remove AI shit

Commit:
chore: make api clear

Commit:
feat: fix issue connected with saving transactions directly

kuzminxyz committed:
Merge branch 'master' into laz/ethexe/3-blocks/network-announce

kuzminxyz committed:
Merge branch 'master' into laz/ethexe/3-blocks/network-announce

kuzminxyz committed:
Merge branch 'laz/ethexe/3-blocks/network-announce' of https://github.com/gear-tech/gear into laz/ethexe/3-blocks/network-announce

kuzminxyz committed:
fix: CI typo check

grishasobol commented on ethexe/common/src/injected.rs:
```diff
@@ -127,6 +128,37 @@ impl InjectedTransaction {
     }
 }
 
+/// Announce injected transaction representation. Contains the signature
+/// over [InjectedTransaction] and its hash.
+///
+/// NOTE: can be constructed only from [`SignedInjectedTransaction`].
+#[derive(Debug, Clone, Encode, Decode, PartialEq, Eq, TypeInfo, Hash)]
+pub struct AnnounceInjectedTransaction {
+    signature: Signature,
+    hash: HashOf<InjectedTransaction>,
+}
```
I completely cannot understand this naming, if it contains injected transaction and signature of injected transaction, then what "announce" is doing here?

ecol-master replied:
Because it transaction, that uses only inside announce, I don't know the correct naming for there.

grishasobol replied:
We can get rid of it, like this for example
```
Announce {
 injected_transactions: Option<H256>,
}

// txs hash points to
type TXS = Vec<(Signature, HashOf<InjectedTransaction>)>;

// (Signature, HashOf<InjectedTransaction>) - is not widely using, so we can leave it as tuple
```

grishasobol commented on ethexe/common/src/primitives.rs:
```diff
@@ -88,9 +89,7 @@ pub struct Announce {
-    // TODO kuzmindev: remove InjectedTransaction from Announce and store only its hashes.
-    // Need to implement `PublicAnnounce` struct which will contain full bodies of injected transactions.
-    pub injected_transactions: Vec<SignedInjectedTransaction>,
+    pub injected_transactions: Vec<AnnounceInjectedTransaction>,
```
`Option<H256>` must be here, see `AnnounceV2`

ecol-master replied:
How `Option<H256>` instead of `Vec<Transaction>`?

ecol-master replied:
Maybe it is ok one single hash, I need to think about it. It differs from initial design

grishasobol replied:
Better single hash, because in case of many injected transactions, vector would lead to unnecessary allocations and work with decoding/encoding/memory copying.

grishasobol commented on ethexe/common/src/primitives.rs:
```diff
@@ -151,6 +150,243 @@ impl ToDigest for Announce {
+    pub fn into_parts(self) -> (Announce, Vec<SignedInjectedTransaction>) {
+        (self.to_announce(), self.injected_transactions)
+    }
+    // #[allow(clippy::result_large_err)]
+    // pub fn try_from_announce(
+    //     announce: Announce,
+    //     injected_transactions: Vec<SignedInjectedTransaction>,
+    // ) -> Result<Self, NetworkAnnounceFromAnnounceError> {
+    //     (announce, injected_transactions).try_into()
+    // }
+
+    // pub fn to_hash(&self) -> HashOf<Announce> {
+    //     Announce::from(self).to_hash()
```
what is this

ecol-master replied:
Sorry, forgot remove in refactoring

grishasobol commented on ethexe/common/src/primitives.rs:
```diff
@@ -151,6 +150,243 @@ impl ToDigest for Announce {
+/// [NetworkAnnounce] is used for transport the [Announce] with [InjectedTransaction] bodies.
+#[cfg_attr(feature = "serde", derive(Hash))]
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq)]
+pub struct NetworkAnnounce {
```
Better to rename this to `FullAnnounce` and use it as data provided for `ethexe-compute` also: ComputeAnnounce(FullAnnounce)

ecol-master replied:
But in the initial design it was `NetworkAnnounce`.

grishasobol replied:
You can leave as Network, but NetworkAnnounce can be used by ethexe-compute also. anyway is not so important

grishasobol commented on ethexe/network/src/db_sync/responses.rs:
```diff
@@ -234,6 +272,14 @@ enum ProcessAnnounceError {
     },
     #[error("announce {hash} not found in database")]
     AnnounceMissing { hash: HashOf<Announce> },
+    #[error("injected transaction {hash} not found in database")]
+    InjectedTransactionMissing { hash: HashOf<InjectedTransaction> },
+    #[error("announce {hash} injected transactions mismatch: {source}")]
+    AnnounceInjectedTransactionsMismatch {
+        hash: HashOf<Announce>,
+        #[source]
+        source: NetworkAnnounceError,
+    },
```
You are mixing here two types of errors. InjectedTransactionMissing and AnnounceInjectedTransactionsMismatch is internal node error, and should be reported at least as "log::error", other errors signals about Request invalidation, so split it to another error type, or create separate RequestInvalidReason

breathx commented:
Will be superseded by the Malachite consensus migration — closing to keep things tidy.
